compilers/concrete-compiler/compiler/lib/Runtime/GPUDFG.cpp

```diff
@@ -1168,20 +1187,28 @@ void memref_bootstrap_lwe_u64_process(Process *p, int32_t loc, int32_t chunk_id,
           cuda_malloc_async(test_vector_idxes_size, s, loc);
       cuda_memcpy_async_to_gpu(test_vector_idxes_gpu, (void *)test_vector_idxes,
                                test_vector_idxes_size, s, loc);
+      // Initialize indexes
+      uint64_t *indexes = (uint64_t*) malloc(num_samples * sizeof(uint64_t));
+      for (uint32_t i = 0; i < num_samples; i++) {
+	indexes[i] = i;
+      }
+      void *indexes_gpu = alloc_and_memcpy_async_to_gpu(indexes, 0, num_samples, loc, s);
+
       int8_t *pbs_buffer = p->dfg->gpus[loc].get_pbs_buffer(
           p->glwe_dim.val, p->poly_size.val, num_samples);
       void *ct0_gpu = d0->device_data;
       void *out_gpu = cuda_malloc_async(data_size, s, loc);
       void *fbsk_gpu = p->ctx.val->get_bsk_gpu(
           p->input_lwe_dim.val, p->poly_size.val, p->level.val, p->glwe_dim.val,
           loc, s, p->sk_index.val);
-      cuda_bootstrap_amortized_lwe_ciphertext_vector_64(
-          s, loc, out_gpu, glwe_ct_gpu, test_vector_idxes_gpu, ct0_gpu,
+      cuda_programmable_bootstrap_amortized_lwe_ciphertext_vector_64(
+          s, loc, out_gpu, indexes_gpu, glwe_ct_gpu, test_vector_idxes_gpu, ct0_gpu, indexes_gpu,
           fbsk_gpu, (int8_t *)pbs_buffer, p->input_lwe_dim.val, p->glwe_dim.val,
-          p->poly_size.val, p->base_log.val, p->level.val, num_samples,
-          lut_indexes.size(), lwe_idx, cuda_get_max_shared_memory(loc));
+          p->poly_size.val, p->base_log.val, p->level.val, num_samples);
       cuda_drop_async(test_vector_idxes_gpu, s, loc);
       cuda_drop_async(glwe_ct_gpu, s, loc);
+      cuda_drop_async(indexes_gpu, s, loc);
+      free(indexes);
       Dependence *dep =
           new Dependence(loc, out, out_gpu, false, false, d0->chunk_id);
       // As streams are not synchronized, we can only free this vector
```

Comment:
I can''t add comment at right line - just 4 lines below is the way to deallocate data after CUDA synch points - it  registers the pointer with the runtime then will deallocate after the next cuda synch.